sidecar/interface.go

```diff
@@ -0,0 +1,230 @@
+package sidecar
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/sha256"
+	"fmt"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/btcsuite/btcutil"
+	"github.com/lightninglabs/pool/order"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// Version is the version of the sidecar ticket format.
+type Version uint8
+
+const (
+	// VersionDefault is the initial version of the ticket format.
+	VersionDefault Version = 0
+)
+
+// State is the state a sidecar ticket currently is in. Each updater of the
+// ticket is responsible for also setting the state correctly after adding their
+// data according to their role.
+type State uint8
+
+const (
+	// StateCreated is the state a ticket is in after it's been first
+	// created and the offer information was added but not signed yet.
+	StateCreated State = 0
+
+	// StateOffered is the state a ticket is in after the offer data was
+	// signed by the sidecar provider.
+	StateOffered State = 1
+
+	// StateRegistered is the state a ticket is in after the recipient has
+	// registered it in their system and added their information.
+	StateRegistered State = 2
+
+	// StateOrdered is the state a ticket is in after the bid order for the
+	// sidecar channel was submitted and the order information was added to
+	// the ticket. The ticket now also contains a signature of the provider
+	// over the order digest.
+	StateOrdered State = 3
+
+	// StateExpectingChannel is the state a ticket is in after it's been
+	// returned to the channel recipient and their node was instructed to
+	// start listening for batch events for it.
+	StateExpectingChannel State = 4
+
+	// StateCompleted is the state a ticket is in after the sidecar channel
+	// was successfully opened and the bid order completed.
+	StateCompleted State = 5
+
+	// StateCanceled is the state a ticket is in after the sidecar channel
+	// bid order was canceled by the taker.
+	StateCanceled State = 6
+)
+
+// Offer is a struct holding the information that a sidecar channel provider is
+// committing to when offering to buy a channel for the recipient. The sidecar
+// channel flow is initiated by the provider creating a ticket and adding its
+// signed offer to it.
+type Offer struct {
```

Comment:
Likely don't need in this version, but I imagine we'll also eventually have the following fields:
  * pubkey of the node that the channel may be opened to 
  * type of channel

Reply:
I'll keep that in mind. With the extensibility of the TLV stream we should be able to easily add those in a later version.
It got me thinking though... Should the individual fields of the sub structs (`Offer`, `Receiver` and so on) also be encoded as TLV records? If I understood the TLV library correctly you can have records of a stream missing and can add new records. But within one record (for example looking at `DOffer`), all the fields that are read with `tlv.DUint64` still need to be present or you run into an EOF.

Could there be an issue with encoding TLV records inside TLV records or should I go ahead and refactor the sub structs to also use records instead of just fields?

Reply:
I went ahead and also used tlv records for the sub structs.
The nice side effect of this is that this also makes it easier to parse the ticket in other languages (for example in JS for a "light" implementation of the sidecar receiver client) as the tlv parsing can just be done recursively.